src/uwtools/rocoto.py

```diff
@@ -143,3 +65,249 @@ def validate_rocoto_xml(input_xml: OptionalPath) -> bool:
+                d = SubElement(e, STR.taskdep)
+                self._set_attrs(d, block)
+            else:
+                raise UWConfigError("Unhandled dependency type %s" % tag)
+
+    def _add_task_envar(self, e: Element, name: str, value: str) -> None:
+        """
+        Add a <envar> element to the <task>.
+
+        :param e: The parent element to add the new element to.
+        :param config: Configuration data for this element.
+        """
+        e = SubElement(e, STR.envar)
+        SubElement(e, STR.name).text = name
+        SubElement(e, STR.value).text = value
+
+    def _add_workflow(self, config: dict) -> None:
+        """
+        Create the root <workflow> element.
+
+        :param config: Configuration data for this element.
+        """
+        config, e = config[STR.workflow], Element(STR.workflow)
+        self._set_attrs(e, config)
+        self._add_workflow_cycledefs(e, config[STR.cycledefs])
+        self._add_workflow_log(e, config[STR.log])
+        self._add_workflow_tasks(e, config[STR.tasks])
+        self._root: Element = e
+
+    def _add_workflow_cycledefs(self, e: Element, config: dict) -> None:
+        """
+        Add <cycledef> element(s) to the <workflow>.
+
+        :param e: The parent element to add the new element to.
+        :param config: Configuration data for this element.
+        """
+        for name, coords in config.items():
+            for coord in coords:
+                SubElement(e, STR.cycledef, group=name).text = coord
+
+    def _add_workflow_log(self, e: Element, logfile: str) -> None:
+        """
+        Add <log> element(s) to the <workflow>.
+
+        :param e: The parent element to add the new element to.
+        :param logfile: The path to the log file.
+        """
+        SubElement(e, STR.log).text = logfile
+
+    def _add_workflow_tasks(self, e: Element, config: dict) -> None:
+        """
+        Add <task> and/or <metatask> element(s) to the <workflow>.
+
+        :param e: The parent element to add the new element to.
+        :param config: Configuration data for these elements.
+        """
+        for key, block in config.items():
+            tag, name = self._tag_name(key)
+            {STR.metatask: self._add_metatask, STR.task: self._add_task}[tag](e, block, name)
+
+    def _insert_doctype(self, xml: str) -> str:
+        """
+        Insert the <!DOCTYPE> block in a rendered XML document.
+
+        :param xml: The XML document rendered as a string.
+        :return: The XML document with the <!DOCTYPE> block inserted.
+        """
+        lines = xml.split("\n")
+        if doctype := self._doctype:
+            lines.insert(1, doctype)
+        return "\n".join(lines)
+
+    def _set_attrs(self, e: Element, config: dict) -> None:
+        """
+        Set attributes on an element.
+
+        :param e: The element to set the attributes on. :config: A config containing the attribute
+            definitions.
```

Comment:
I haven't seen this syntax on the params section (that I remember).

Reply:
It's a typo, will fix.